dask/array/core.py

```diff
@@ -1141,7 +1149,7 @@ def __getitem__(self, index):
         dsk, chunks = slice_array(out, self.name, self.chunks, index2)
 
         graph = HighLevelGraph.from_collections(out, dsk, dependencies=[self])
-        return Array(graph, out, chunks, dtype=self.dtype)
+        return Array(graph, out, chunks, meta=self._meta.astype(self.dtype))
```

Comment:
Hrm, what if slicing changes the type of the array?

For example, I could imagine that slicing some types of sparse matrices might result in other types of sparse matrices, or possibly even numpy arrays.  Ideally we transform `meta` as we perform the operation, rather than just changing around the dtype.  In this way our dask array's meta type will track what happens in the lazy chunks.

So for an operation like `__getitem__` we might make a new index that was just like the input, but had a bunch of elements that were safe to apply against an empty array.

```python
In [17]: x = np.empty((0, 0, 0))

In [18]: index2 = (slice(0, 0), [], None, slice(0, 0))  # like index, but empty

In [19]: x[index2]
Out[19]: array([], shape=(0, 0, 1, 0), dtype=float64)
```

However I'm not sure how best to handle integer elements in the index.  I'm not sure that there is an empty-safe way to handle this.

Reply:
Yes, I think that will need some more careful thinking. We were aware of this kind of situation, unless I misunderstood, this case is analogous to the case of a Dask array containing multiple arrays of different types, which was mentioned by you in #2977:

> The plan I've proposed above doesn't support the mixed-array case (where we have a mix of dense and sparse arrays in the same dask array for example), but I suggest that we start there. I think that it will be a good and productive first step.

Reply:
I think that this concern is independent of mixed chunks.  

Lets say that some CSR matrix type exists, and I have a dask array full of them.

When I do something like `x[0, :]` I pull out a row.  It could be that the return type of that object is no longer a CSR matrix, but some other form (maybe COO).  The output still has a uniform chunk type, it's just a new type.

Reply:
I think I get it now. I'll try to think how to handle it.

Reply:
The best I could do so far was:

```python
meta = self._meta[index].astype(self.dtype)
meta.shape = (0,) * len(chunks)
```

That second line probably suffices to keep the shape as we need it, unless the result should be a 0d array (for such case, I don't have a solution yet). But if the index could change the type of the matrix, isn't there a case that the actual array size will matter and _not_ change it if the matrix is empty?

To be honest, at this point I'm starting to ask myself if this is really a good design. There are so many ifs and buts that I think it may immediately render the code unmaintainable.

Reply:
Oh, I forgot to mention that the code above works only for slicing, indexing would need to be a separate case.

Reply:
> To be honest, at this point I'm starting to ask myself if this is really a good design. 

It's a good question to ask.  And thank you for going through the experiment that helps us to understand whether or not this is the right approach.

> There are so many ifs and buts that I think it may immediately render the code unmaintainable.

We've gone through this process before both with dtype and with shape.  It's *definitely* a challenge.  

My guess is that we'll be able to find a solution for most cases.  For the more difficult cases there are people watching this PR that have deeper NumPy experience than you or I do (people like @shoyer @jakirkham @hameerabbasi and @TomAugspurger ) and they can help once we identify the difficult issues.  

At this point we've identified that slicing is going to be difficult.  Lets move on to other operations before diving too deeply here, and maybe one of the people above can offer a suggestion based on the description above.

Reply:
> We've gone through this process before both with dtype and with shape. It's definitely a challenge.

I understand this, and I'm not even worried about NumPy, but answering the question of what could 3rd-party libraries do to their own types in every possible case may be untenable. I'm thinking that waiting for `empty_like()` to be available may be the right choice, since we could then just recreate `_meta` with the right size every time something may have changed, rather than spending so much time in handling all possible cases.

> At this point we've identified that slicing is going to be difficult. Lets move on to other operations before diving too deeply here, and maybe one of the people above can offer a suggestion based on the description above.

I ended up on slicing, because fixing the issue below (in `partial_reduce()`) depends on slicing. At some point things fail because the `_meta` array has the wrong `shape`. I suspect this will be a requirement for other pieces of the code as well, since it's one of the most basic operations.

Reply:
@mrocklin I pushed some new commits, one of them attempts to handle slicing of `_meta`. This has been capable of handling all the cases that previously failed in existing tests, but there's always a chance that I've missed cases that I couldn't test for.

Reply:
> However I'm not sure how best to handle integer elements in the index. I'm not sure that there is an empty-safe way to handle this.

There is, just prepend `Ellipsis` to the list of indexes.

```python
>>> import numpy as np
>>> np.float64(0)[..., None, None]
array([[0.]])
```

And if you want to guarantee that the output is an array, even if 0-D, postpend with `Ellipsis`:
```python
>>> np.array([[0]])[0, 0, ...]
array(0)
```